llvm/include/llvm/ADT/ArrayRef.h

```diff
@@ -149,6 +149,13 @@ namespace llvm {
                  * = nullptr)
         : Data(Vec.data()), Length(Vec.size()) {}
 
+    /// Construct an ArrayRef<T> from iterator_range<U*>. This uses SFINAE
+    /// to ensure that this is only used for iterator ranges over plain pointer
+    /// iterators.
+    template <typename U, typename = std::enable_if_t<std::is_same_v<U *, T *>>>
```

Comment:
Sorry about the thinko, but why is template parameter U needed if `std::is_same_iv<U *, T *>`? Doesn't that imply that U = T?

Reply:
Yep, this should use `is_convertible` I think, to support creating `ArrayRef<const X>` from `iterator_range<X *>`, also added a test

Reply:
Not 100% sure if is_convertible is the best thing to do here: maybe `is_same_v<U, remove_const_t<T>>` is better?

Reply:
Unfortunately `is_same_v<U, remove_const_t<T>>` isn't sufficient to support pointer cases with different const-ness directly .

I updated the `is_convertible_v` check to add another level of indirection, which should forbids construction a ArrayRef for a non-pointer base from an iterator of derived*, the only case that was previously allowed. This is similar to how `std::vector` and `SmallVectorImpl` are handled above. Added a few more static asserts to the test